diff --git a/config.sub b/config.sub
index 38f3d037a78..dcab58066bd 100755
--- a/config.sub
+++ b/config.sub
@@ -1749,7 +1749,7 @@ case $os in
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx* | zephyr* \
-	     | fiwix* )
+	     | fiwix* | piggyOS* )
 		;;
 	# This one is extra strict with allowed versions
 	sco3.2v2 | sco3.2v[4-9]* | sco5v6*)
diff --git a/gcc/config.gcc b/gcc/config.gcc
index b5eda046033..4b34f5a5a89 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -239,7 +239,7 @@ default_use_cxa_atexit=no
 default_gnu_indirect_function=no
 target_gtfiles=
 need_64bit_isa=
-native_system_header_dir=/usr/include
+native_system_header_dir=/initrd/include
 target_type_format_char='@'
 
 # Don't carry these over build->host->target.  Please.
@@ -677,6 +677,12 @@ x86_cpus="generic intel"
 
 # Common parts for widely ported systems.
 case ${target} in
+*-*-piggyOS*)
+  gas=yes
+  gnu_ld=yes
+  default_use_cxa_atexit=yes
+  use_gcc_stdint=provide
+  ;;
 *-*-darwin*)
   tmake_file="t-darwin "
   tm_file="${tm_file} darwin.h"
@@ -1113,6 +1119,10 @@ case ${target} in
 esac
 
 case ${target} in
+i[34567]86-*-piggyOS*)
+    default_gnu_indirect_function=yes
+    tm_file="${tm_file} i386/unix.h i386/att.h dbxelf.h elfos.h glibc-stdint.h i386/i386elf.h piggyOS.h i386/piggyOS.h"
+    ;;
 aarch64*-*-elf | aarch64*-*-fuchsia* | aarch64*-*-rtems*)
 	tm_file="${tm_file} elfos.h newlib-stdint.h"
 	tm_file="${tm_file} aarch64/aarch64-elf.h aarch64/aarch64-errata.h aarch64/aarch64-elf-raw.h"
diff --git a/gcc/config/i386/piggyOS.h b/gcc/config/i386/piggyOS.h
new file mode 100644
index 0000000000000000000000000000000000000000..53a4b8e93b74b4808a4bfed91c4d5558217c584a
--- /dev/null
+++ b/gcc/config/i386/piggyOS.h
@@ -0,0 +1,7 @@
+/* Ensure that we are using the SIZE_TYPE indicated by SysV */
+#undef SIZE_TYPE
+#define SIZE_TYPE    (TARGET_64BIT ? "long unsigned int" : "unsigned int")
+
+/* Ensure that ptrdiff_t matches the actual pointer size */
+#undef PTRDIFF_TYPE
+#define PTRDIFF_TYPE    (TARGET_64BIT ? "long int" : "int")
diff --git a/gcc/config/piggyOS.h b/gcc/config/piggyOS.h
new file mode 100644
index 0000000000000000000000000000000000000000..17551aaa1a07e6c0b7365f98899375122ba12529
--- /dev/null
+++ b/gcc/config/piggyOS.h
@@ -0,0 +1,28 @@
+#undef TARGET_PIGGYOS
+#define TARGET_PIGGYOS 1
+ 
+#undef LIB_SPEC
+#define LIB_SPEC "-lc" /* link against C standard library */
+
+#undef STANDARD_STARTFILE_PREFIX
+#define STANDARD_STARTFILE_PREFIX "/initrd/lib/"
+
+/* Files that are linked before user code.
+   The %s tells GCC to look for these files in the library directory. */
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC "crt0.o%s crti.o%s crtbegin.o%s"
+ 
+/* Files that are linked after user code. */
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+ 
+/* Additional predefined macros. */
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()      \
+  do {                                \
+    builtin_define ("__piggyOS");      \
+    builtin_define ("__unix__");      \
+    builtin_assert ("system=piggyOS");   \
+    builtin_assert ("system=unix");   \
+    builtin_assert ("system=posix");   \
+  } while(0);
